bam_sort.c

```diff
@@ -1611,13 +1633,71 @@ static int change_SO(bam_hdr_t *h, const char *so)
 }
 
 // Function to compare reads and determine which one is < the other
-static inline int bam1_lt(const bam1_p a, const bam1_p b)
+// Handle sort-by-pos and sort-by-name. Used as the secondary sort in bam1_lt_by_tag, if reads are equivalent by tag.
+static inline int bam1_lt_core(const bam1_p a, const bam1_p b)
 {
     if (g_is_by_qname) {
         int t = strnum_cmp(bam_get_qname(a), bam_get_qname(b));
         return (t < 0 || (t == 0 && (a->core.flag&0xc0) < (b->core.flag&0xc0)));
-    } else return (((uint64_t)a->core.tid<<32|(a->core.pos+1)<<1|bam_is_rev(a)) < ((uint64_t)b->core.tid<<32|(b->core.pos+1)<<1|bam_is_rev(b)));
+    } else {
+        return (((uint64_t)a->core.tid<<32|(a->core.pos+1)<<1|bam_is_rev(a)) < ((uint64_t)b->core.tid<<32|(b->core.pos+1)<<1|bam_is_rev(b)));
+    }
+}
+
+// Sort record by tag, using pos or read name as a secondary key if tags are identical. Reads not carrying the tag sort first. 
+// Tags are first sorted by the type character (in case the types differ), or by the appropriate comparator for that type if they agree.
+static inline int bam1_lt_by_tag(const bam1_p a, const bam1_p b) 
+{
+    uint8_t* aux_a = bam_aux_get(a, g_sort_tag);
+    uint8_t* aux_b = bam_aux_get(b, g_sort_tag);
+
+    if (aux_a == 0 && aux_b > 0) {
+        return 1;
+    } else if (aux_a > 0 && aux_b == 0) {
+        return 0;
+    } else if (aux_a == 0 && aux_b == 0) {
+        return bam1_lt_core(a,b);
+    } 
+
+    if (*aux_a < *aux_b) {
```

Comment:
This doesn't quite work, as for integer tags `*aux_a` and `*aux_b` can have different types (say c and i), but still be comparable.  You might want to cover the integer case before this (and do we want to allow comparison of floats and integers?)